plugins/zynaddsubfx/zynaddsubfx/src/Misc/QtXmlWrapper.cpp

```diff
@@ -193,9 +216,11 @@ int QtXmlWrapper::dosavefile(const char *filename,
                            int compression,
                            const char *xmldata) const
 {
+   int fd = fdopenFromUtf8(filename, QIODevice::WriteOnly);
```

Comment:
Why not use the same method here that's used in `RemoteVstPlugin.cpp`? We can move the `toWString` method and the `F_OPEN` defines to a common header and include that in both files.

Reply:
I think `F_OPEN` macro is an ugly workaround for `fopen`. I can do that if needed, but I don't like the macro approach.

Reply:
@PhysSong You can make it a function instead.

Reply:
Something tricky about that: `"r"` vs `L"r"` or something like that. That's why I chose macro approach.

Reply:
@PhysSong Sounds like you need to convert the mode string as well. So something like this:

```c++
std::FILE* openFile(std::string filename, std::string mode)
{
#ifdef LMMS_BUILD_WIN32
	return _wfopen(toWString(filename), toWString(mode));
#else
	return fopen(filename.data(), mode.data());
#endif
}
```

Reply:
Looks fine for me, except missing `data()` for WIN32.

Reply:
Can you elaborate on that?

Reply:
I mean `return _wfopen(toWString(filename).data(), toWString(mode).data());`.

Reply:
Of course, sorry. I thought you were telling me `data()` isn't available for WIN32 and started getting angry at MinGW. :)

Reply:
I don't really know which approach will be better here. Both approach seem to have not much overheads. Qt provides a nice API for this, but `_wfopen` looks somewhat straightforward.
Alternatively, `CreateFileW` + [`_open_osfhandle`](https://msdn.microsoft.com/en-us/library/bdts1c9x.aspx)(what `QFile` internally does) may work for getting file descriptor.

Reply:
Neither is really better than the other, but they do the same thing so I think they should be consolidated into a single function. The Qt solution is not an option for this because it's not available in `RemoteVstPlugin`.

Reply:
> but they do the same thing
They *almost* do. On Windows, `QFile` holds native file handle and converts it to a C file handle when requested.
`_wfopen` creates a `FILE` which holds a C handle to the file. That's why I used the former here, but I can switch to the latter if needed.

Reply:
Surely `F_OPEN` and `fdopenFromUtf8` fulfil the same task, which is opening a C `FILE` given an UTF8-encoded file name, right? Even though they don't work the same way internally, they "do the same thing" as in "they solve the same problem". From that point of view, this is duplicate code and there should only be one function for it.

Reply:
Then where should I move `F_OPEN` and related stuffs? I can't find a good place for it.
I was not using that in Zyn because it(more precisely `zlib`) needs `fdopen`-able handle, not a `FILE` pointer. However, getting(and duplicating) file handle is not hard so I can easily add it.

Reply:
Good question, I can't find a good place either. Maybe just create a new header? Something like `lmmsio.h` (inspired by `stdio.h`), although that name is very similar to https://lmms.io/.

Reply:
I also used `QFile` for getting file descriptors which will be used by `libsndfile`. Do you think it's worth unifying everything with `F_OPEN` based macros/functions?

Reply:
I think either `lmms_io.h` or `IoHelper.h` will be fine. I'll address this point soon.

Reply:
Or how about putting them in `RemotePlugin.h`? I can't use-cases of UTF-8 string filename except in remote plugins.

Reply:
Is it fine for our Zyn to include LMMS' headers at all? I think this will be fine in `stable-1.2`, but I'm not sure about `master`'s submodule. It would be the last concern for me...

Reply:
I know it doesn't make sense to build https://github.com/LMMS/zynaddsubfx as a standalone, of course. However, I think it's possible to check if it's building with LMMS or not, though not really needed.